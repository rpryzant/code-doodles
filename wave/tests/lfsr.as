;;; An implementation of a LFSR-based random number generator.
;;; Input is the initial seed.  Prints all value in orbit
;;; (c) 2010 duane a. bailey
	b	main
	.equ	perline,16
ntaps:	.data	4
taps:	.data	7,5,4,3

main:	swi	#SysGetNum
	mov	r1,r0
	mov	r2,#perline
	bl	setSeed
	swi	#SysPutNum
	sub	r2,r2,#1
	mov	r0,#'\ 
	swi	#SysPutChar
_loop:	bl	nextInt
	swi	#SysPutNum
	stu	r0,[sp,#-1]
	subs	r2,r2,#1
	moveq	r2,#perline
	moveq	r0,#'\n
	swieq	#SysPutChar
	movne	r0,#'\ 
	swine	#SysPutChar
	ldu	r0,[sp,#1]
	cmp	r0,r1
	bne	_loop
	cmp	r2,#perline
	movne	r0,#'\n
	swine	#SysPutChar
	swi	#SysHalt

seed:	.data	0
setSeed: stu	r1,[sp,#-1]
	str	r0,seed
	ldu	r1,[sp,#1]
	mov	pc,lr

nextInt:stm	sp,#0x1fe
	ldr	r1,ntaps	; number of taps
	adr	r2,taps		; tap pointer
	mov	r3,#0		; sum of taps
	ldr	r4,seed		; grab seed
_loop:	subs	r1,r1,#1	; look at the next tap
	blt	_done
	ldr	r6,[r2,r1,lsl#0]	; grab the current tap
	mov	r5,#1
	and	r5,r5,r4,lsr r6	; shift seed and mask low bit
	eor	r3,r3,r5	; add into running sum
	b	_loop
_done:	mvn	r8,#0
	mov	r7,#31
	sub	r7,r7,r6	; find the width of the generator
	mov	r8,r8,lsr r7
	mov	r4,r4,lsl #1
	orr	r4,r4,r3
	and	r0,r4,r8
	str	r0,seed
	ldm	sp,#0x1fe
	mov	pc,lr
	