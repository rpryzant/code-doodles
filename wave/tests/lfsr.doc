Implements a linear feedback shift register on 8 bits.
Prints the full orbit of the LFSR based on a seed found on input.
