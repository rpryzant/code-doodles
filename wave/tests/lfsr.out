13 27 54 108 216 177 99 199 143 30 60 121 243 231 206 156
57 115 230 204 152 49 98 197 139 22 45 90 180 105 210 164
72 145 34 69 138 20 41 82 165 74 149 42 84 169 83 167
78 157 59 119 238 221 187 118 236 217 179 103 207 158 61 123
247 239 223 191 126 253 250 244 233 211 166 76 153 51 102 205
154 53 106 212 168 81 163 70 140 24 48 96 193 131 7 14
29 58 117 234 213 170 85 171 87 175 95 190 124 249 242 229
202 148 40 80 161 66 132 9 19 39 79 159 63 127 255 254
252 248 240 225 194 133 11 23 47 94 188 120 241 227 198 141
26 52 104 208 160 64 128 1 2 4 8 17 35 71 142 28
56 113 226 196 137 18 37 75 151 46 92 184 112 224 192 129
3 6 12 25 50 100 201 146 36 73 147 38 77 155 55 110
220 185 114 228 200 144 32 65 130 5 10 21 43 86 173 91
182 109 218 181 107 214 172 89 178 101 203 150 44 88 176 97
195 135 15 31 62 125 251 246 237 219 183 111 222 189 122 245
235 215 174 93 186 116 232 209 162 68 136 16 33 67 134 13
